snap/snapenv/snapenv.go

```diff
@@ -22,16 +22,31 @@ package snapenv
 import (
 	"fmt"
 	"os"
+	"os/user"
+	"strings"
 
 	"github.com/snapcore/snapd/arch"
 	"github.com/snapcore/snapd/snap"
 )
 
-// Basic returns the app-level environment variables for a snap.
+// ExecEnv returns the environment that is important for the later stages
+// of execution (like SNAP_REVISION that snap-exec requires to work)
+func ExecEnv(info *snap.Info) []string {
+	// merge environment and the snap environment, note that the
+	// snap environment overrides pre-existing env entries
+	env := envMap(os.Environ())
+	snapEnv := envMap(snapEnv(info))
+	for k, v := range snapEnv {
+		env[k] = v
+	}
+	return envFromMap(env)
+}
+
+// basicEnv returns the app-level environment variables for a snap.
 // Despite this being a bit snap-specific, this is in helpers.go because it's
 // used by so many other modules, we run into circular dependencies if it's
 // somewhere more reasonable like the snappy module.
-func Basic(info *snap.Info) []string {
+func basicEnv(info *snap.Info) []string {
```

Comment:
note that here you're building a `[]string` and then parsing it out into a map, which seems silly (hence why my suggestion (in code, rather obscure, yadda) of making Basic just return a map (and making User take the map and add things to it)).